src/game.cpp

```diff
@@ -9862,7 +9862,55 @@ void game::reload()
         reload( item_loc );
 
     } else {
-        reload( -1 );
+        // As a special streamlined activity, hitting reload repeatedly should:
+        // Reload wielded gun
+        // First reload a magazine if necessary.
+        // Then load said magazine into gun.
+        // Reload magazines that are compatible with the current gun.
+        // Reload other guns in inventory.
+        // Reload misc magazines in inventory.
+        std::vector<item_location> reloadables = u.find_reloadables();
+        std::sort( reloadables.begin(), reloadables.end(),
+        [this]( const item_location & a, const item_location & b ) {
+            const item *ap = a.get_item();
+            const item *bp = b.get_item();
+            // Current wielded weapon comes first.
+            if( ap == &this->u.weapon ) {
+                return true;
+            }
+            if( bp == &this->u.weapon ) {
+                return false;
+            }
+            // Second sort by afiliation with wielded gun
+            const std::set<itype_id> compatible_magazines = this->u.weapon.magazine_compatible();
+            if( compatible_magazines.count( ap->typeId() ) > 0 &&
+                compatible_magazines.count( bp->typeId() ) == 0 ) {
+                return true;
+            }
+            if( compatible_magazines.count( bp->typeId() ) > 0 &&
+                compatible_magazines.count( ap->typeId() ) == 0 ) {
+                return false;
+            }
+            // Third sort by gun vs magazine,
+            if( ap->is_gun() && !bp->is_gun() ) {
+                return true;
+            }
+            if( bp->is_gun() && !ap->is_gun() ) {
+                return false;
```

Comment:
I'd suggest a bit less repetitive comparison:

```c++
const bool mag_ap = compatible_magazines.count( ap->typeId() ) > 0;
const bool mag_bp = compatible_magazines.count( bp->typeId() ) > 0;

if( mag_ap != mag_bp ) {
    return mag_ap;
}

if( ap->is_gun() != bp->is_gun() ) {
    return ap->is_gun();
}
```

Reply:
Thank you so much, I kept thinking there was a better option like this, but it didn't come to mind.

Reply:
You're very welcome. I've once encountered this approach and it serves me well since then.